src/Octoshift/Services/MultipartUploaderService.cs

```diff
@@ -0,0 +1,129 @@
+            {
+                nextUrl = await UploadPart(buffer, bytesRead, nextUrl.ToString());
+            }
+
+            // 3. Complete the upload
+            await CompleteUpload(nextUrl.ToString());
+
+            return $"gei://archive/{guid}";
+        }
+        catch (Exception ex)
+        {
+            throw new OctoshiftCliException("Failed during multipart upload.", ex);
+        }
+    }
+
+    private async Task<IEnumerable<KeyValuePair<string, IEnumerable<string>>>> StartUpload(string uploadUrl, string archiveName, long contentSize)
+    {
+        var body = new
+        {
+            content_type = "application/octet-stream",
+            name = archiveName,
+            size = contentSize
+        };
+
+        try
+        {
+            // Post with the expectation of receiving both response content and headers
+            var response = await _client.PostWithFullResponseAsync(uploadUrl, body);
+            return response.ResponseHeaders.ToList();
+        }
+        catch (Exception ex)
+        {
+            throw new OctoshiftCliException("Failed to start upload.", ex);
+        }
+    }
+
+    private async Task<Uri> UploadPart(byte[] body, int bytesRead, string nextUrl)
+    {
+        var content = new ByteArrayContent(body, 0, bytesRead);
+        content.Headers.ContentType = new("application/octet-stream");
+
+        try
+        {
+            // Make the PATCH request and retrieve headers
+            var patchResponse = await _client.PatchWithFullResponseAsync(nextUrl, content);
+            var headers = patchResponse.ResponseHeaders.ToList();
+
+            // Retrieve the next URL from the response headers
+            return GetNextUrl(headers) ?? throw new OctoshiftCliException("Failed to retrieve the next URL for the upload part.");
+        }
+        catch (Exception ex)
+        {
+            throw new OctoshiftCliException("Failed to upload part.", ex);
+        }
+    }
+
+    private async Task CompleteUpload(string lastUrl)
+    {
+        var content = new StringContent(string.Empty);
+        content.Headers.ContentType = new("application/octet-stream");
+
+        try
+        {
+            await _client.PutAsync(lastUrl, content);
+        }
+        catch (Exception ex)
+        {
+            throw new OctoshiftCliException("Failed to complete upload.", ex);
+        }
+    }
+
+    private Uri GetNextUrl(IEnumerable<KeyValuePair<string, IEnumerable<string>>> headers)
+    {
+        // Use FirstOrDefault to safely handle missing Location headers
+        var locationHeader = headers.First(header => header.Key.Equals("Location", StringComparison.OrdinalIgnoreCase));
+
+        if (!string.IsNullOrEmpty(locationHeader.Key))
+        {
+            var locationValue = locationHeader.Value.FirstOrDefault();
+            if (locationValue.HasValue())
+            {
+                return new Uri(new Uri(_base_url), locationValue);
+            }
+        }
+        return null;
```

Comment:
I I would rewrite this method like the following:
```suggestion
        var locationHeader = headers.First(header => header.Key.Equals("Location", StringComparison.OrdinalIgnoreCase));

        if (locationHeader.Key.IsNullOrWhiteSpace() || locationHeader.Value.FirstOrDefault().IsNullOrWhiteSpace())
        {
            throw new OctoshiftCliException("Failed to retrieve the next URL for the upload part.");
        }

        return new Uri(new Uri(BASE_URL), locationHeader.Value.First());
```

Reply:
Is there something wrong or inefficient with my logic?

Reply:
The only main thing was to check for the `null` values and throwing an exception in the method. Aside from that my suggestion also simplified the logic a bit.